Enable reporting of unhandled exceptions obtained from CurrentDomain.UnhandledException on Android

evgeny-pol opened the pull request:
We can't report unhandled exception obtained from CurrentDomain.UnhandledException callback thread because of JNI errors arising on attempt to call any Java class:
`E/zygote: JNI ERROR (app bug): accessed stale Local 0x9  (index 0 in a table of size 0)`
`A/zygote: java_vm_ext.cc:534] JNI DETECTED ERROR IN APPLICATION: use of deleted local reference 0x9`

They do not occur if we send exception reports from the main thread

evgeny-pol committed:
report unhandled exceptions using main thread

evgeny-pol committed:
fix StopUnhandledExceptionsCoroutine

MatkovIvan commented on Assets/AppCenter/AppCenterBehavior.cs:
```diff
@@ -8,28 +8,27 @@
-
     public AppCenterSettings settings;
 
+    public static AppCenterBehavior Instance { get; private set; }
```
Do not expose it to public

MatkovIvan commented on Assets/AppCenter/Plugins/AppCenterSDK/Crashes/Shared/Crashes.cs:
```diff
@@ -106,13 +124,67 @@ public static Models.ErrorReport LastSessionCrashReport()
+        {
+            if (_unhandledExceptionsCoroutine == null && AppCenterBehavior.Instance != null)
+            {
+                _unhandledExceptionsCoroutine = AppCenterBehavior.Instance.StartCoroutine(SendUnhandledExceptionReports());
```
Use `UnityCoroutineHelper`

MatkovIvan commented on Assets/AppCenter/Plugins/AppCenterSDK/Crashes/Shared/Crashes.cs:
```diff
@@ -106,13 +124,67 @@ public static Models.ErrorReport LastSessionCrashReport()
+                    {
+                        var exception = _unhandledExceptions.Dequeue();
+                        var exceptionWrapper = CreateWrapperException(exception);
+                        CrashesInternal.TrackException(exceptionWrapper.GetRawObject());
```
Move it out of `lock`

evgeny-pol committed:
Unhandled exceptions reporting update

subscribe to app logs only when needed
do not make AppCenterBehavior.instance field public
use UnityCoroutineHelper to start coroutines
do not lock object on every frame
stop coroutine using flag instead of stopping it using StopCoroutine
move some operations out of lock

MatkovIvan commented on Assets/AppCenter/Plugins/AppCenterSDK/Crashes/Shared/Crashes.cs:
```diff
@@ -56,16 +68,26 @@ public static void OnHandleLog(string logString, string stackTrace, LogType type
             var exception = args.ExceptionObject as Exception;
             if (exception != null)
             {
+                Debug.Log("Unhandled exception: " + exception.ToString());
```
`AppCenterLog`
